### PR TITLE
Improve handling of allOf and add tests

### PR DESCRIPTION
**Pull request by matthchr**
- Perform deduplication and add some tests.

### CODE REVIEW

File: hack/generator/pkg/jsonast/jsonast.go [hunk 1]
```diff
@@ -590,6 +615,7 @@ func anyOfHandler(ctx context.Context, scanner *SchemaScanner, schema *gojsonsch
 	}
 
 	// return all possibilities...
+	klog.Errorf("Unhandled anyOf type: %v\n", schema.Ref.GetUrl())
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/jsonast/jsonast.go:618]: 👍

File: hack/generator/pkg/jsonast/jsonast.go [hunk 2]
```diff
@@ -603,7 +629,7 @@ func arrayHandler(ctx context.Context, scanner *SchemaScanner, schema *gojsonsch
 
 	if len(schema.ItemsChildren) == 0 {
 		// there is no type to the elements, so we must assume interface{}
-		klog.Warning("Interface assumption unproven\n")
+		klog.Warningf("Interface assumption unproven for %v\n", schema.Ref.GetUrl())
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/jsonast/jsonast.go:632]: Including the URL gives us a lever to find the problem. 👍

File: hack/generator/pkg/jsonast/testdata/AllOf/AllOf_generates_wrapper_type.json [hunk 3]
```diff
@@ -0,0 +1,50 @@
+{
+    "$comment": "Check that the AllOf handler correctly generates a union type. Note that this will also generate the standalone types Foo, Bar, and Baz which might actually be a bug (see discussion: " ,
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/jsonast/testdata/AllOf/AllOf_generates_wrapper_type.json:2]: Is this comment truncated?

File: hack/generator/pkg/jsonast/testdata/UnreferencedDefinitions/unreferenced_definitions_are_not_generated.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+            "type": "object",
+            "properties": {
+                "enabled": {
+                    "type": "boolean"
+                }
+            },
+            "required": ["enabled"]
+        }
+    }
+}
```
Comment (theunrepentantgeek) [inline hack/generator/pkg/jsonast/testdata/UnreferencedDefinitions/unreferenced_definitions_are_not_generated.json:40]: Missing EOLN

File: hack/generator/pkg/jsonast/testdata/AllOf/AllOf_generates_wrapper_type.json (hunk 3 above)
Comment (matthchr) [reply hack/generator/pkg/jsonast/testdata/AllOf/AllOf_generates_wrapper_type.json:2]: Yes

File: hack/generator/pkg/jsonast/jsonast.go [hunk 5]
```diff
@@ -442,32 +442,57 @@ func allOfHandler(ctx context.Context, scanner *SchemaScanner, schema *gojsonsch
...
-			// if it's a reference to a defined struct, embed it inside:
-			fields = append(fields, astmodel.NewEmbeddedStructDefinition(s))
+			if def, ok := scanner.findTypeDefinition(concreteType); ok {
+				var err error
+				fields, err = handleType(fields, def.Type())
+				if err != nil {
+					return nil, err
+				}
+			} else {
+				return nil, fmt.Errorf("couldn't find struct for definition: %v", concreteType)
```
Comment (Porges) [inline hack/generator/pkg/jsonast/jsonast.go:477]: this doesn't actually check if it's a struct type; it could be an enum

**Pull request merged**